cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -2758,6 +2773,438 @@ struct GroupedOneFactory {
   InputType argument_type;
 };
 
+// ----------------------------------------------------------------------
+// List implementation
+
+template <typename Type, typename Enable = void>
+struct GroupedListImpl final : public GroupedAggregator {
+  using CType = typename TypeTraits<Type>::CType;
+  using GetSet = GroupedValueTraits<Type>;
+
+  Status Init(ExecContext* ctx, const std::vector<ValueDescr>&,
+              const FunctionOptions* options) override {
+    ctx_ = ctx;
+    has_nulls_ = false;
+    // out_type_ initialized by GroupedListInit
+    values_ = TypedBufferBuilder<CType>(ctx_->memory_pool());
+    groups_ = TypedBufferBuilder<uint32_t>(ctx_->memory_pool());
+    values_bitmap_ = TypedBufferBuilder<bool>(ctx_->memory_pool());
+    return Status::OK();
+  }
+
+  Status Resize(int64_t new_num_groups) override {
+    num_groups_ = new_num_groups;
+    return Status::OK();
+  }
+
+  Status Consume(const ExecBatch& batch) override {
+    const auto values_array_data = batch[0].array();
+    int64_t num_values = values_array_data->length;
+
+    const auto groups_array_data = batch[1].array();
+    DCHECK_EQ(groups_array_data->offset, 0);
+    const auto* groups = groups_array_data->GetValues<uint32_t>(1, 0);
+    RETURN_NOT_OK(groups_.Append(groups, num_values));
+
+    int64_t offset = values_array_data->offset;
+    const uint8_t* values = values_array_data->buffers[1]->data();
+    RETURN_NOT_OK(GetSet::AppendBuffers(&values_, values, offset, num_values));
+    //    RETURN_NOT_OK(values_.Resize(num_args_ + num_values));
+    //    CopyDataUtils<Type>::CopyData(*batch[0].type(), *values_array_data, offset,
+    //                                  values_.bytes_builder()->mutable_data(),
+    //                                  values_.length(), num_values);
+    //    values_.bytes_builder()->UnsafeAdvance(num_values * sizeof(CType));
```

Comment:
This is what using `CopyData` might look like (be it then refactoring it into a helper). Also, `CopyData` does not support decimal types.

Reply:
DecimalType is just a special case of FixedSizeBinaryType as far as we're concerned here.